dwst/scripts/particles.js

```diff
@@ -20,7 +20,37 @@ const specialChars = [
   '\\',
 ];
 
-export class InvalidParticles extends Error { }
+export class InvalidParticles extends Error {
+  constructor(remainder, expected, expectedLengthOverride) {
+    super();
+    this.expression = null;
+    this.errorPosition = null;
+    this.remainder = remainder;
+    this.expected = expected;
+    if (typeof expectedLengthOverride === 'undefined') {
+      this.expected_length = expected.length;
+    } else {
+      this.expected_length = expectedLengthOverride;
+    }
+    this.got = remainder.substring(0, this.expected_length);
+  }
+}
+
+export function invalidToMlog(invalidParticle) {
+  return [
+    'Invalid template.',
+    invalidParticle.expression,
+    `${' '.repeat(invalidParticle.errorPosition)}^`,
```

Comment:
maybe move the code to `const padding = ...`

Reply:
done.